qiskit_addon_sqd/fermion.py

```diff
@@ -66,6 +79,351 @@ def load(cls, filename):
+                ord=np.inf,
+            )
+            < occupancies_tol
+        ):
+            break
+        current_result = best_result_in_batch
+        current_occupancies = current_result.orbital_occupancies
+
+        # Carry over bitstrings with large CI weight
+        sci_state = current_result.sci_state
+        flattened = sci_state.amplitudes.reshape(-1)
+        absolute_vals = np.abs(flattened)
+        indices = np.argsort(absolute_vals)
+        carryover_index = np.searchsorted(absolute_vals, carryover_threshold, sorter=indices)
+        carryover_indices = indices[carryover_index:]
+        _, n_strings_b = sci_state.amplitudes.shape
+        alpha_indices, beta_indices = np.divmod(carryover_indices, n_strings_b)
+        carryover_strings_a = sci_state.ci_strs_a[alpha_indices]
+        carryover_strings_b = sci_state.ci_strs_b[beta_indices]
+        if symmetrize_spin:
+            carryover_strings_a = carryover_strings_b = np.union1d(
+                carryover_strings_a, carryover_strings_b
+            )
+
+    # best_result is not None because there must have been at least one iteration
+    return cast(SCIResult, best_result)
+
+
+def solve_sci_batch(
+    ci_strings: list[tuple[np.ndarray, np.ndarray]],
+    one_body_tensor: np.ndarray,
+    two_body_tensor: np.ndarray,
+    norb: int,
+    nelec: tuple[int, int],
+    *,
+    spin_sq: float | None = None,
+    **kwargs,
+) -> list[SCIResult]:
+    """Diagonalize Hamiltonian in subspaces.
+
+    Args:
+        ci_strings: List of pairs (strings_a, strings_b) of arrays of spin-alpha CI
+            strings and spin-beta CI strings whose Cartesian product give the basis of
+            the subspace in which to perform a diagonalization.
+        one_body_tensor: The one-body tensor of the Hamiltonian.
+        two_body_tensor: The two-body tensor of the Hamiltonian.
+        norb: The number of spatial orbitals.
+        nelec: The numbers of alpha and beta electrons.
+        spin_sq: Target value for the total spin squared for the ground state.
+            If ``None``, no spin will be imposed.
+        **kwargs: Keyword arguments to pass to `pyscf.fci.selected_ci.kernel_fixed_space <https://pyscf.org/pyscf_api_docs/pyscf.fci.html#pyscf.fci.selected_ci.kernel_fixed_space>`_
+
+    Returns:
+        The results of the diagonalizations in the subspaces given by ci_strings.
+    """
+    return [
+        solve_sci(
+            ci_strs,
+            one_body_tensor,
+            two_body_tensor,
+            norb=norb,
+            nelec=nelec,
+            spin_sq=spin_sq,
+            **kwargs,
+        )
+        for ci_strs in ci_strings
+    ]
+
+
+def solve_sci(
+    ci_strings: tuple[np.ndarray, np.ndarray],
+    one_body_tensor: np.ndarray,
+    two_body_tensor: np.ndarray,
+    norb: int,
+    nelec: tuple[int, int],
+    *,
+    spin_sq: float | None = None,
+    **kwargs,
+) -> SCIResult:
+    """Diagonalize Hamiltonian in subspace.
```

Comment:
"... in subspace defined by `ci_strs`." might sound a little more complete. I know `ci_strings` description covers it, but I think it'd be nice here too